### PR TITLE
Add `SchemaDefinitions` and support JSON Schema `$ref`

### PR DESCRIPTION
**Pull request by tomchristie**
Refs #19

For example:

```python
import typesystem

definitions = typesystem.SchemaDefinitions()

class Album(typesystem.Schema, definitions=definitions):
    title = typesystem.String(max_length=100)
    release_date = typesystem.Date()
    artist = typesystem.Reference('Artist')

class Artist(typesystem.Schema, definitions=definitions):
    name = typesystem.String(max_length=100)
```

TODO:

- [ ] Support setting the namespace in all possible nested attribute types. Eg. `additional_items=...`
- [x] Raise an error if the string representation is used within a schema and no namespace is set.
- [x] Raise an error if a duplicate name is used within a namespace.
- [x] Dumping namespaces and refs to JSON schema.
- [x] Loading refs from JSON schema.

We *may* want more involved stuff like nested namespaces E.g. mapping onto `{"$ref": "#/definitions/auth/User"}` but let's just leave that all well alone to start with.

To consider: "Namespace" or "Definitions"?

### CODE REVIEW

**Comment by wbolster** [comment]
see also: https://github.com/encode/typesystem/issues/19#issuecomment-470960705

**Comment by tomchristie** [comment]
@wbolster Updated in line with your suggestion. Yup that's better.

File: typesystem/fields.py [hunk 1]
```diff
@@ -655,16 +655,28 @@ def __init__(self, **kwargs: typing.Any) -> None:
...
         super().__init__(**kwargs)
         self.schema = schema
+        self.namespace = namespace
 
     def validate(self, value: typing.Any, *, strict: bool = False) -> typing.Any:
+        if isinstance(self.schema, str):
+            assert (
+                self.namespace is not None
+            ), "String-references can only be used within a schema namespace."
+            schema = self.namespace[self.schema]
```
Comment (wbolster) [inline typesystem/fields.py]: you may want to consider to assign to `self.schema` as well to only do this one to ‘settle’ the names on first use.

File: typesystem/schemas.py [hunk 2]
```diff
@@ -4,11 +4,16 @@
...
 from typesystem.fields import Field, Object
+from typesystem.namespaces import SchemaNamespace, set_namespace
 
 
 class SchemaMetaclass(ABCMeta):
     def __new__(
-        cls: type, name: str, bases: typing.Sequence[type], attrs: dict
+        cls: type,
+        name: str,
+        bases: typing.Sequence[type],
```
Comment (wbolster) [inline typesystem/schemas.py]: not 100% sure, but this should be `typing.Sequence[typing.Type]` i think

File: typesystem/fields.py (hunk 1 above)
Comment (tomchristie) [reply typesystem/fields.py]: Yes probably a good plan. Less scope for behavioural oddities then.

File: typesystem/namespaces.py [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
 from collections.abc import MutableMapping
 
-from typesystem.schemas import Schema
+from typesystem.fields import Array, Field, Nested, Object
 
 
 class SchemaNamespace(MutableMapping):
     def __init__(self) -> None:
-        self._namespace = {}  # type: typing.Dict[str, typing.Type[Schema]]
+        self._namespace = {}  # type: dict
```
Comment (wbolster) [inline typesystem/namespaces.py]: is there a reason to keep using comments instead of real annotation syntax?

File: typesystem/schemas.py (hunk 2 above)
Comment (tomchristie) [reply typesystem/schemas.py]: I think it’s right as it stands. Type would also work, but makes more sense if we wanted specificity. E.g. can use `list` or `typing.List` for an annotation, but you’d usually only use the later for list-of-x

File: typesystem/namespaces.py (hunk 3 above)
Comment (tomchristie) [reply typesystem/namespaces.py]: Don’t recall what versions that starts being supported in?

Comment (tomchristie) [reply typesystem/namespaces.py]: Oh right, 3.6. Yeah could defo consider a PR changing that across the board.

File: typesystem/schemas.py (hunk 2 above)
Comment (wbolster) [reply typesystem/schemas.py]: ah thanks for explaining. i am still new to `typing`.

**Comment by wbolster** [comment]
> @wbolster Updated in line with your suggestion. Yup that's better.

wow i thought using a `namespace=...` metaclass arg would be very complex, but it seems it's actually not at all. i like how it looks, much nicer than subclassing `ns.Schema` instead of `Schema`! :+1:

File: typesystem/namespaces.py (hunk 3 above)
Comment (wbolster) [reply typesystem/namespaces.py]: see #50

**Pull request merged**